eloquent-javascript-exercises/chap4/chap4.js

```diff
@@ -0,0 +1,107 @@
+//write program range function to take an optional third argument that indicates the “step” value used when building the array. If no step is given, the elements go up by increments of one, corresponding to the old behavior.
+function range(start, end, step = 1) {
+    let numbers = [];
+    if (step > 0) {
+        for (let i = start; i <= end; i += step) {
+            numbers.push(i);
+        }
+    }
+    else {
+        for (let i = start; i >= end; i += step) {
+            numbers.push(i);
+        }
+    }
+    return numbers;
+}
+
+
+
+//write a sum function that takes an array of numbers and returns the sum of these numbers.
+
+
+function sum(array) {
+    return array.reduce((total, num) => total + num);
+}
+
+
+//Arrays have a reverse method that changes the array by inverting the order in which its elements appear. For this exercise, write two functions, reverseArray and reverseArrayInPlace. The first, reverseArray, takes an array as argument and produces a new array that has the same elements in the inverse order. The second, reverseArrayInPlace, does what the reverse method does: it modifies the array given as argument by reversing its elements. Neither may use the standard reverse method.
+function reverseArray(array) {
+    let newArray = [];
+    while (array.length) {
+        newArray.push(array.pop());
+    }
+    return newArray;
+}
+
+
+function reverseArrayInPlace(arr) {
+    function swapArrayItems(index1, index2) {
+        if (index1 == index2) {
+            return arr;
+        }
+        let temp = arr[index1];
+        arr[index1] = arr[index2]
+        arr[index2] = temp;
+        return swapArrayItems(index1 + 1, index2 - 1);
```

Comment:
alternative for swapping
[arr[index1], arr[index2]] = [arr[index2], arr[index1]]